toolchain/lower/mangler.cpp

```diff
@@ -0,0 +1,91 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/lower/mangler.h"
+
+#include "toolchain/base/kind_switch.h"
+#include "toolchain/sem_ir/entry_point.h"
+
+namespace Carbon::Lower {
+
+auto Mangler::MangleInverseQualifiedNameScope(bool first_name_component,
+                                              llvm::raw_ostream& os,
+                                              SemIR::NameScopeId name_scope_id)
+    -> void {
+  auto& sem_ir = file_context_.sem_ir();
+  auto names = sem_ir.names();
+  auto types = sem_ir.types();
+  while (name_scope_id.is_valid()) {
+    const auto& parent = sem_ir.name_scopes().Get(name_scope_id);
+    if (parent.inst_id == SemIR::InstId::PackageNamespace) {
+      break;
+    }
+    if (!first_name_component) {
+      os << '.';
+    }
+    first_name_component = false;
+    CARBON_KIND_SWITCH(sem_ir.insts().Get(parent.inst_id)) {
+      case CARBON_KIND(SemIR::ImplDecl impl_decl): {
+        const auto& impl = sem_ir.impls().Get(impl_decl.impl_id);
+        if (auto opt_class_self =
+                types.TryGetAs<SemIR::ClassType>(impl.self_id)) {
+          MangleInverseQualifiedNameScope(
```

Comment:
We're avoiding recursive functions in order to avoid stack depth issues. Can you build an iterative approach for this?

Reply:
Sure thing - think I've got that sorted, and basically meant figuring out mangling qualified interface names in too, so that's added/working.